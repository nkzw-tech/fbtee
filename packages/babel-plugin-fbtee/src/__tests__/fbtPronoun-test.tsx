import { describe, expect, it } from '@jest/globals';
import {
  jsCodeFbtCallSerializer,
  snapshotTransform,
  transform,
  withFbtImportStatement,
} from './FbtTestUtil.tsx';

expect.addSnapshotSerializer(jsCodeFbtCallSerializer);

describe('fbt pronoun support', () => {
  it('"capitalize" option accepts boolean literal true', () => {
    expect(
      snapshotTransform(
        withFbtImportStatement(
          `var x = fbt(
            fbt.pronoun('possessive', gender, {capitalize: true}) +
              ' birthday is today.',
            'Capitalized possessive pronoun',
          );`,
        ),
      ),
    ).toMatchSnapshot();
  });

  it('Should throw when using non-Boolean option value', () => {
    // Note: Using StringLiteral '"true"' instead of BooleanLiteral 'true'.
    const input = withFbtImportStatement(
      `var x = fbt(
        'Today is ' +
          fbt.pronoun('possessive', gender, {human: 'true'}) +
          ' a happy birthday.',
        'Expect error exception',
      );`,
    );
    expect(() => transform(input)).toThrow(
      "Expected boolean value instead of 'true' (string)",
    );
  });

  it('Should throw when using non-Boolean option value in a template', () => {
    // Note: Using StringLiteral '"true"' instead of BooleanLiteral 'true'.
    const input = withFbtImportStatement(
      `var x = fbt(
        \`Today is \${fbt.pronoun('possessive', gender, {
          human: 'true',
        })} a happy birthday.\`,
        'Expect error exception',
      );`,
    );
    expect(() => transform(input)).toThrow(
      "Expected boolean value instead of 'true' (string)",
    );
  });

  it('Should throw when using unknown "usage" value', () => {
    // Note: Using "possession" instead of "possessive".
    const input = withFbtImportStatement(
      `var x = fbt(
        'Today is ' +
          fbt.pronoun('possession', gender, {human: false}) +
          ' a happy birthday.',
        'Expect error exception',
      );`,
    );
    expect(() => transform(input)).toThrow(
      '`usage`, the first argument of fbt.pronoun() - Expected value to be ' +
        "one of [object, possessive, reflexive, subject] but we got 'possession' (string) instead",
    );
  });

  it('Should elide false "human" option from fbt.pronoun()', () => {
    expect(
      snapshotTransform(
        // I.e. Wish them a happy birthday.
        withFbtImportStatement(
          `var x = fbt(
            'Wish ' +
              fbt.pronoun('object', gender, {human: true}) +
              ' a happy birthday.',
            'Elided false option',
          );`,
        ),
      ),
    ).toMatchSnapshot();
  });
});
