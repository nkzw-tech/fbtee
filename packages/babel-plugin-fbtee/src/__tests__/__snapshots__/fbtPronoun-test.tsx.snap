// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fbt pronoun support "capitalize" option accepts boolean literal true 1`] = `
import { fbt } from "fbtee";
var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      m: [null],
      t: {
        1: {
          desc: "Capitalized possessive pronoun",
          text: "Her birthday is today.",
        },
        2: {
          desc: "Capitalized possessive pronoun",
          text: "His birthday is today.",
        },
        "*": {
          desc: "Capitalized possessive pronoun",
          text: "Their birthday is today.",
        },
      },
    },
    project: "",
  } /* __FBT__ end */,
  [fbt._pronoun(1, gender)]
);

`;

exports[`fbt pronoun support Should elide false "human" option from fbt.pronoun() 1`] = `
import { fbt } from "fbtee";
var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      m: [null],
      t: {
        1: { desc: "Elided false option", text: "Wish her a happy birthday." },
        2: { desc: "Elided false option", text: "Wish him a happy birthday." },
        "*": {
          desc: "Elided false option",
          text: "Wish them a happy birthday.",
        },
      },
    },
    project: "",
  } /* __FBT__ end */,
  [
    fbt._pronoun(0, gender, {
      human: 1,
    }),
  ]
);

`;
